pkg/streamingpromql/operators/aggregation.go

```diff
@@ -154,64 +153,72 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
 
 		thisSeriesGroup := a.remainingInnerSeriesToGroup[0]
 		a.remainingInnerSeriesToGroup = a.remainingInnerSeriesToGroup[1:]
+		err = a.seriesIntoGroup(s, thisSeriesGroup, steps, start, interval)
+		if err != nil {
+			return types.InstantVectorSeriesData{}, err
+		}
+	}
 
-		if len(s.Floats) > 0 && thisSeriesGroup.floatSums == nil {
-			// First series for this group, populate it.
+	// Construct the group and return it
+	seriesData, err := a.constructSeriesData(thisGroup, start, interval)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
 
-			thisSeriesGroup.floatSums, err = a.Pool.GetFloatSlice(steps)
-			if err != nil {
-				return types.InstantVectorSeriesData{}, err
-			}
+	a.Pool.PutFloatSlice(thisGroup.floatSums)
+	a.Pool.PutBoolSlice(thisGroup.floatPresent)
+	a.Pool.PutHistogramPointerSlice(thisGroup.histogramSums)
+	thisGroup.floatSums = nil
+	thisGroup.floatPresent = nil
+	thisGroup.histogramSums = nil
+	thisGroup.histogramPointCount = 0
```

Comment:
I think all of this could move into `constructSeriesData` - then this method (`NextSeries`) becomes very high-level, and all the details are hidden away inside the other methods.